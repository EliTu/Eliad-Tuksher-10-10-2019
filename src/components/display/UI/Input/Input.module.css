@import './../../../../variables.css';

.InputStyles {
    width: 98%;
    box-sizing: border-box;
    font-family: inherit;
}

.InputStyles input,
textarea {
    background-color: rgba(241, 238, 238, 0);
    color: var(--color-main-dark);
    font-size: 1.2rem;
    width: 100%;
    height: 100%;
    padding: 0.3rem 1rem;
    margin: 0;
    display: inline-block;
    transition: all 0.5s;
    outline: none;
}

.validationStyles {}

.invalidStyle {}

.errorMessageStyle {}