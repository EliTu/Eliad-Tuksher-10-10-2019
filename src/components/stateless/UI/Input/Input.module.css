@import './../../../../variables.css';

.InputStyles {
    width: 100%;
    box-sizing: border-box;
    font-family: inherit;
}

.InputStyles input,
textarea {
    background-color: rgba(241, 238, 238, 0);
    color: var(--color-main-dark);
    font-size: 1.2rem;
    width: 100%;
    height: 100%;
    border: 2px solid;
    border-top-left-radius: 8px;
    border-bottom-left-radius: 8px;
    border-color: var(--color-main-dark);
    padding: 0.3rem 2rem;
    margin: 0;
    display: inline-block;
    transition: all 0.5s;
}

.validationStyles {}

.invalidStyle {}

.errorMessageStyle {}