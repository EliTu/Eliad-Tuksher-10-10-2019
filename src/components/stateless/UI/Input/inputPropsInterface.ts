// The Input component prop type TS interface
export interface IInputProps {
	elementType: string;
	elementConfig: {
		type: string;
		placeholder: string;
		label: string;
		options?: [{ value: string; displayValue: string; disabled: boolean }];
		min?: number;
		max?: number;
	};
	value: string | number;
	isFocused: boolean;
	handleChange?: (event: object) => void;
}
