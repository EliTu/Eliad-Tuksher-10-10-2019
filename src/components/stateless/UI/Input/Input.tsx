import React, { useRef, useEffect } from 'react';
import styles from './Input.module.css';
import { IInputProps } from './inputPropsInterface';

const Input: React.FC<IInputProps> = ({
	elementType,
	elementConfig,
	value,
	validation,
	isFocused,
	handleChange,
	handleEnterPress,
}) => {
	const { InputStyles, invalidStyle, ValidStyle, errorMessageStyle } = styles;

	let inputElement = null;

	let errorMessageElement: JSX.Element = (
		<p className={errorMessageStyle}>{validation.errorMessage}</p>
	);

	// Listen to keyboard enter click to submit form:
	const enterPressCallback = (event: any, func: (e: Event) => void) => {
		if (event.key === 'Enter') func(event);
	};

	// Focus the first input field upon component mount
	const focusRef: React.RefObject<any> = useRef();

	useEffect(() => {
		if (isFocused) focusRef.current.focus();
	}, [isFocused]);

	// Set validation styles:
	let validationStyles: string = '';
	validationStyles =
		!validation.valid && validation.hasUserInput
			? invalidStyle
			: validation.valid && validation.hasUserInput && value !== ''
			? ValidStyle
			: '';

	switch (elementType) {
		case 'input':
			inputElement = (
				<input
					ref={focusRef}
					className={validationStyles}
					{...elementConfig}
					value={value}
					onChange={handleChange}
					onKeyPress={event =>
						enterPressCallback(event, handleEnterPress)
					}
					data-test="input-test"
				/>
			);
			break;

		case 'textarea':
			inputElement = (
				<textarea
					{...elementConfig}
					value={value}
					onChange={handleChange}
					onKeyPress={event =>
						enterPressCallback(event, handleEnterPress)
					}
					data-test="textarea-test"
				/>
			);
			break;

		case 'select':
			inputElement = (
				<select
					value={value}
					onChange={handleChange}
					data-test="select-test"
				>
					{elementConfig.options &&
						elementConfig.options.map(option => (
							<option
								value={option.value}
								key={option.value}
								disabled={option.disabled}
							>
								{option.displayValue}
							</option>
						))}
				</select>
			);
			break;

		default:
			inputElement = (
				<input
					{...elementConfig}
					value={value}
					// className={!validation.valid ? InvalidStyle : null}
					onKeyPress={event =>
						enterPressCallback(event, handleEnterPress)
					}
				/>
			);
	}
	return (
		<div className={InputStyles}>
			<label>{elementConfig.label}</label>
			{inputElement}
			{errorMessageElement}
		</div>
	);
};

export default Input;
