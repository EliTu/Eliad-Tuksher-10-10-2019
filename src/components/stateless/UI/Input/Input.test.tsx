import React from 'react';
import Input from './Input';
import { IInputProps } from '../Input/inputPropsInterface';
import { shallow, ShallowWrapper } from 'enzyme';
import { findByTestAttr } from './../../../../utilities/test-utilities/findByTestAttr';

const onChange: () => void = jest.fn();
const onEnterPress: () => void = jest.fn();

const setComponentProps = (
	elementType: string,
	configType: string = 'text',
	value: string = 'abc',
	isValid: boolean = false,
	hasInput: boolean = true,
	options: object[]
) => {
	const props: IInputProps = {
		elementType: `${elementType}`,
		elementConfig: {
			type: `${configType}`,
			placeholder: 'abc',
			label: 'abc',
			options: [
				{
					value: 'some value',
					displayValue: 'Some Value',
					key: '1',
				},
				{
					value: 'another value',
					displayValue: 'Another Value',
					key: '2',
				},
			],
		},
		value: `${value}`,
		validation: {
			required: true,
			hasUserInput: `${hasInput}`,
			valid: `${isValid}`,
			errorMessage: 'abc',
		},
		focused: true,
		handleChange: () => {},
		handleEnterPress: () => {},
		data: 'abc',
	};
	return props;
};

const setComponent = (
	props: IInputProps,
	changeFn = () => {},
	enterFn = () => {}
) => {
	const component: ShallowWrapper = shallow(
		<Input {...props} handleChange={changeFn} handleEnterPress={enterFn} />
	);
	return component;
};

describe('Input component', () => {
	describe('Component (not testing the input fields)', () => {
		it('should render an Input component withot errors', () => {
			let props = setComponentProps('input', 'text', 'abc', true);
			let component = setComponent(props);
			expect(component).toMatchSnapshot();
			expect(component).toBeTruthy();
			expect(component).toHaveClassName('InputStyles');
			expect(component.length).toBe(1);
			expect(component.length).not.toBe(2);
			expect(component).toHaveClassName('InputStyles');
		});
	});

	describe('Type: input="text" test', () => {
		let props = setComponentProps('input');
		let component = setComponent(props);
		let input = findByTestAttr(component, 'input-test');

		it('should render the component without errors', () => {
			expect(component).toMatchSnapshot();
			expect(input.length).toBe(1);
		});

		it('should not render if elementType is not input ', () => {
			props = setComponentProps('password');
			component = setComponent(props);
			input = findByTestAttr(component, 'input-test');

			expect(component).toMatchSnapshot();
			expect(input.length).toBe(0);
		});

		it('should call the onChange event callback function on user input', () => {
			props = setComponentProps('input', null, '');
			component = setComponent(props, onChange);
			input = findByTestAttr(component, 'input-test');

			input.simulate('change', {
				target: { value: 'abc' },
			});

			expect(onChange).toHaveBeenCalled();
		});

		it('should call the onKeyPress event callback function on user press enter', () => {
			props = setComponentProps('input');
			component = setComponent(props, null, onEnterPress);
			input = findByTestAttr(component, 'input-test');

			input.simulate('keypress', { key: 'Enter' });

			expect(onEnterPress).toHaveBeenCalled();
		});
	});

	describe('Type: input="textarea" test', () => {
		let props = setComponentProps('textarea', '');
		let component = setComponent(props);
		let textarea = findByTestAttr(component, 'textarea-test');
		it('should render without errors', () => {
			expect(component).toMatchSnapshot();
			expect(textarea.length).toBe(1);
			expect(textarea.length).not.toBe(2);
		});

		it('should call the onKeyPress event callback function on user press enter', () => {
			component = setComponent(props, null, onEnterPress);

			textarea.simulate('keypress', { key: 'Enter' });

			expect(onEnterPress).toHaveBeenCalled();
		});
	});

	describe('Type: input="select" test', () => {
		let props = setComponentProps('select');
		let component = setComponent(props);
		let select = findByTestAttr(component, 'select-test');

		it('should render without errors', () => {
			expect(component).toMatchSnapshot();
			expect(select.length).toBe(1);
			expect(select.find('option').length).toBe(2);
		});

		it('should call the onChange event callback function on user change option', () => {
			select.simulate('change', { target: { value: 'efg' } });
			expect(onChange).toHaveBeenCalled();
		});
	});
});
