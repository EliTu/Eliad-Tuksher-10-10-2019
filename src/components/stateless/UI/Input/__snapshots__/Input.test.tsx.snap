// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Component (not testing the input fields) should render an Input component withot errors 1`] = `
<div
  className="InputStyles"
>
  <label>
    abc
  </label>
  <input
    className="ValidStyle"
    data-test="input-test"
    label="abc"
    onChange={[Function]}
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Some Value",
          "key": "1",
          "value": "some value",
        },
        Object {
          "displayValue": "Another Value",
          "key": "2",
          "value": "another value",
        },
      ]
    }
    placeholder="abc"
    type="text"
    value="abc"
  />
  <p
    className="errorMessageStyle"
  >
    abc
  </p>
</div>
`;

exports[`Input component Type: input="select" test should render without errors 1`] = `
<div
  className="InputStyles"
>
  <label>
    abc
  </label>
  <select
    data-test="select-test"
    onChange={[Function]}
    value="abc"
  >
    <option
      key="some value"
      value="some value"
    >
      Some Value
    </option>
    <option
      key="another value"
      value="another value"
    >
      Another Value
    </option>
  </select>
  <p
    className="errorMessageStyle"
  >
    abc
  </p>
</div>
`;

exports[`Input component Type: input="text" test should not render if elementType is not input  1`] = `
<div
  className="InputStyles"
>
  <label>
    abc
  </label>
  <input
    label="abc"
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Some Value",
          "key": "1",
          "value": "some value",
        },
        Object {
          "displayValue": "Another Value",
          "key": "2",
          "value": "another value",
        },
      ]
    }
    placeholder="abc"
    type="text"
    value="abc"
  />
  <p
    className="errorMessageStyle"
  >
    abc
  </p>
</div>
`;

exports[`Input component Type: input="text" test should render the component without errors 1`] = `
<div
  className="InputStyles"
>
  <label>
    abc
  </label>
  <input
    className="ValidStyle"
    data-test="input-test"
    label="abc"
    onChange={[Function]}
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Some Value",
          "key": "1",
          "value": "some value",
        },
        Object {
          "displayValue": "Another Value",
          "key": "2",
          "value": "another value",
        },
      ]
    }
    placeholder="abc"
    type="text"
    value="abc"
  />
  <p
    className="errorMessageStyle"
  >
    abc
  </p>
</div>
`;

exports[`Input component Type: input="textarea" test should render without errors 1`] = `
<div
  className="InputStyles"
>
  <label>
    abc
  </label>
  <textarea
    data-test="textarea-test"
    label="abc"
    onChange={[Function]}
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Some Value",
          "key": "1",
          "value": "some value",
        },
        Object {
          "displayValue": "Another Value",
          "key": "2",
          "value": "another value",
        },
      ]
    }
    placeholder="abc"
    type=""
    value="abc"
  />
  <p
    className="errorMessageStyle"
  >
    abc
  </p>
</div>
`;
